hangman/src/game_state.rs

```diff
@@ -0,0 +1,189 @@
+        self.word_so_far.guesses.clone()
+    }
+
+    pub fn get_obfuscated_word(&self) -> String {
+        format!("{}", self.word_so_far)
+    }
+
+    fn score(&self, new_word_so_far: &WordSoFar, guess: char) -> RoundOutcome {
+        let is_duplicate = self.word_so_far.has_seen(&guess);
+        let is_in_word = self.word_so_far.target_word.contains(guess);
+        let num_blanks_remaining = new_word_so_far.get_remaining_letters();
+
+        match (
+            is_duplicate,
+            is_in_word,
+            num_blanks_remaining,
+            self.num_wrong_remaining,
+        ) {
+            (_, _, _, 0) => RoundOutcome::Lose,
+            (_, _, 0, _) => RoundOutcome::Win,
+            (true, _, _, _) => RoundOutcome::Duplicate,
+            (_, false, _, _) => RoundOutcome::Miss,
+            _ => RoundOutcome::Hit,
+        }
+    }
+
+    pub fn guess(&self, guess: char) -> GameState {
+        let new_word_so_far = self.word_so_far.add_guess(guess);
+        let round_outcome = self.score(&new_word_so_far, guess);
+        let num_wrong_remaining = self.num_wrong_remaining
+            - if round_outcome == RoundOutcome::Miss {
+                1
+            } else {
+                0
+            };
+
+        GameState {
+            word_so_far: new_word_so_far,
+            round_outcome,
+            num_wrong_remaining,
+            last_guess: guess,
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum RoundOutcome {
+    /// Discovered a new letter
+    Hit,
+    /// Found a letter not in the word
+    Miss,
+    /// Already guessed this letter
+    Duplicate,
+    /// Successfully found the word
+    Win,
+    /// Failed to find the word
+    Lose,
+}
+
+impl fmt::Display for GameState {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self.round_outcome {
+            RoundOutcome::Hit => write!(f, ""),
+            RoundOutcome::Miss => write!(
+                f,
+                "{}",
+                format!(
+                    "'{}' in not in the word. You have {} more incorrect guesses.",
+                    self.last_guess, self.num_wrong_remaining
+                )
+                .red()
+            ),
+            RoundOutcome::Duplicate => write!(
+                f,
+                "{}",
+                format!("You've already guessed '{}'.", self.last_guess).red()
+            ),
+            RoundOutcome::Win => write!(
+                f,
+                "{}",
+                format!("You win! The word was '{}'.", self.word_so_far.target_word).green()
+            ),
+            RoundOutcome::Lose => write!(
+                f,
+                "{}",
+                format!("You lose! The word was '{}'.", self.word_so_far.target_word).red()
+            ),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum HangmanError {
```

Comment:
So these aren't actually errors for the hangman game, they're parse errors for your UI. I would probably keep this in main and rename it `ParseError`. (or put all your terminal-based UI code in its own module and have main be like 1 line long.)